### PR TITLE
feat: enable burns on migration

### LINKED ISSUE #7552: Migration UI
### Tasks: 

- https://github.com/kodadot/nft-gallery/issues/7560
- https://github.com/kodadot/nft-gallery/issues/7561
- https://github.com/kodadot/nft-gallery/issues/7562
- https://github.com/kodadot/nft-gallery/issues/7563

image

# Abstract
- design available in handoff file: https://www.figma.com/file/JXNtryaNYhY0XJNArzvPQx/Handoff?type=design&node-id=4218%3A2062&mode=design&t=VM0ypGEgcdUaSoT1-1
image
- this ui should enable migration capabalities
- consists of 4 pages - home page, overview, signing, congratulations message
- many parts similar to the admin panel, signing modals, modals in general, drop page (cards).
- as in the admin panel issue, it's quite big, please use the developer option in Figma - you easily see the padding etc. 

### Tasks: 

- https://github.com/kodadot/nft-gallery/issues/7560
- https://github.com/kodadot/nft-gallery/issues/7561
- https://github.com/kodadot/nft-gallery/issues/7562
- https://github.com/kodadot/nft-gallery/issues/7563

Now for the design: 

------

------

------

# Home page

## Top part
image
- introduces into migration, shows important parts - like the route etc.

### source and destination
- same as in teleport v2 (#7363) 
- gives the option to choose whether migrating from Polkadot assethub or Kusama asset hub. The default is polkadot. 
image
- The source is Kusama, or if its more clearer maybe rmrk2, It is up to the devs 

### Faq
- I have decided to put faq on the top there
- migration is pretty hard to understand for new users. This will hopefully ensure the users are not confused about what is happening. 
image

- states are on the left: 
image
- please note that the height will be taken from the card with the most text - other cards will inherit the height (that means cards are consistent and won't jump around) 
- if you would like to add anything as an explainer to the cards, feel free to comment. 

## Selection
- under the full-width divider, there is a contextual message with the account you are connected with, and then there is a little tooltip section. 
- there is also a state for when you are not connected. 
image
- tooltip copy for that as well. 
image

- the selection is filled with drop landing page-like cards. 
image

- cards have a hover effect, and the migrate button itself has one as well. 
image

- you can try it yourself in the prototype next to the card component. 

### Ready for Migration
- Here, you will be able to view collections that you own.
- >As the creator or owner of a collection, you can initiate migration here. Keep in mind only items owned by you can be migrated at this stage. You'll need to pre-sign items owned by others to allow individual migrations.
- if there is no collection to be migrated from this part, this part is simply hidden. 

### Waiting for your action
image

- here are items from collections you don't own, but the collection was migrated and initiated - giving you the option to migrate as well. 
- >Displayed here are individual items that you own from collections that have been migrated. Ensure you're up-to-date by migrating these pieces right away. Please note you can only start the migration once the collection owner initiates it.

### migration not possible
- pretty sure there will be some collections that you won't be able to migrate. For this we have the last part under divider. 
image

------

------

------

small divider 🍡 

# Review
- you will get there by clicking any button with label "migrate...." on the cards 
- the UI divides here to migration of the mentioned collection or only the items inside. 

## Migrate collection
image

- top part review - sign (same as in the admin panel) 
- contextual message ( connected with)
- then there is migrate collection 
- there is picture of the collection, name and message that the name will stay the same.
- divider, and now for the harder part. 
- there is row that shows "items available for migration" 
- little message: 
   - >Please note that at this stage, only items you own can be migrated. For other items, you will need to pre-sign them to allow others to follow your migration initiative.
- If I understand correctly, you cannot migrate items you don't own. Therefore, you pre-sign for others and give them access to follow your migration once they decide to do so. 
- is that correct @vikiival ? 
- then there is route explained
image

- divider and then costs, those are also similar to the #7241 
- collapsable fee breakdown, divided into network fee and then existential deposit, and then kodadot fee (if there will be any..) 
- then a row of estimated total fees. 

### bottom part 
- here is acknowledgment before you proceed, its irreversible and you should know it. 
- disabled button till you don't check the checkbox 

## Migrate items
image
- missing available items and the info part. 
- only 2 items max are shown and then there is text of how many items are left. 
- >+ 10 more items from the same collection 

------

------

------

# Signing 🚥 
- new part, will be the same as with auto teleport (but in modal) 
- the reason everything is on its own page and not in the small modal is simply the space and no way of accidentally closing it and losing the progress. It's quite a lot of transactions if you decide to migrate 1k collection. 
image

- it's a step signing.
- here are possible events of the small sub-part 
image

- the parent section changes based on whether its in progress, its done, or inactive, changes between three states. 
image image image

- the grey 1/4 circle is a loading wheel, please match the style as much as possible. 
- The same icons are used for the subsections. Every step is in this case one signed transaction. 
- let me give an example. 
image

- the contextual message is in this case what is changing. 
- when you cancel signing, you have the option to retry (it could say failed or canceled) 
image

- after you finish all of this long signing, you can click on the button to finish (it's not automatic because in this signing there are also transaction links (as you can see above) 

------

------

------

# Congratulations 
- not only you have made it to the end of this issue, but the user successfully signed all the transactions and migrated. 

image

- graphics, copy, and buttons to view the collection and to share on Twitter. Copy for the tweet can be done ad hoc. 
- I will create the graphics for dark mode as well, don't worry! 

### End 
- if you have any questions, feel free to ask

### LINKED ISSUE #8474: Migration info resizing when opening sidebar [Firefox only]
https://github.com/kodadot/nft-gallery/assets/36627808/fffb4973-fafb-4b5f-98c9-4cebd683ac6f

### PR DESCRIPTION
**Pull request by preschian**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

- [x] Closes #7552
- [x] Fix #8474

image

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 66ae295</samp>

This pull request adds the functionality to migrate NFTs and collections from one chain to another, using the `minimark` library to display the migration steps. It also fixes the placeholder image path for the `MediaItem` component in both the `components/base/MediaItem.vue` and the `libs/ui/src/components/MediaItem/MediaItem.vue` files.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 66ae295</samp>

> _Sing, O Muse, of the daring deeds of the kodadot heroes_
> _Who burned their NFTs and collections on the source chain_
> _And migrated them to a new and better realm of art and glory_
> _With `transaction` and `useAsyncQuery` as their faithful guides_

### CODE REVIEW

File: components/migrate/steps/SignLoader3.vue [hunk 1]
```diff
@@ -1,5 +1,27 @@
 <template>
   <div class="mb-5">
+    <div class="flex items-center mb-4">
+      <div class="mr-5">
+        <NeoIcon
+          v-if="fromNfts?.length === nextNfts?.length"
+          v-bind="iconSuccess"
+          class="fa-2x" />
+        <NeoIcon v-else v-bind="iconIdle" class="fa-2x" />
```
Comment (roiLeo) [inline components/migrate/steps/SignLoader3.vue:9]: size prop?

**Review by roiLeo — APPROVED** [review]
code lgtm

**Comment by prury** [comment]
@preschian hit me up on dms so we can do some joint migrations and normal ones

**Comment by prury** [comment]
### Round 1
Source collection(only had one item): 



https://deploy-preview-8437--koda-canary.netlify.app/ksm/collection/7EA1DCF47E98A25067-RAI

### whole process(really smooth): 

https://github.com/kodadot/nft-gallery/assets/36627808/9fe6c0e3-3892-400f-be8f-5a0cf4399acc

Transactions: 

Collection creation: 
https://assethub-kusama.subscan.io/tx/0xae765c671ada808805b6d36e10fa337fa0f5342ceea5736e5bd3ea3e24a3d331

Migration:
https://assethub-kusama.subscan.io/tx/0xc76f6fa945c2bf0bd58d48d72201b184b155184741bd959156673becf10601ce

NFT burning: 
https://kusama.subscan.io/tx/0xbc2f187ec5644176f88928e9fdd53edd6eb3c005b7e37e200e1025e1ace504a2

Destroy Collection - i think this one is wrong, because it was made on Assethub instead of Kusama ❌ 
https://assethub-kusama.subscan.io/tx/0xe1053104dedbd5c67e0cff61f67e470ea5c63cd375de63b6c4a8577a784a640f

File: components/migrate/steps/SignLoader3.vue (hunk 1 above)
Comment (preschian) [reply components/migrate/steps/SignLoader3.vue:9]: somehow "large" prop didn't work, and "medium" still `smaller` than `fa-2x`

**Comment by preschian** [comment]
> Destroy Collection - i think this one is wrong, because it was made on Assethub instead of Kusama ❌
> [assethub-kusama.subscan.io/tx/0xe1053104dedbd5c67e0cff61f67e470ea5c63cd375de63b6c4a8577a784a640f](https://assethub-kusama.subscan.io/tx/0xe1053104dedbd5c67e0cff61f67e470ea5c63cd375de63b6c4a8577a784a640f)

Fixed, please note: even if we burn the collection, it still shows up in our profile. Need to add `burned` on the indexer side

On AssetHub, we can check the `burned` on the query for collection, but not on the rmrk

image
.

**Comment by prury** [comment]
> 
> Fixed, please note: even if we burn the collection, it still shows up in our profile. Need to add `burned` on the indexer side
> 

is there a open issue for that? so we can implement in the future

**Comment by prury** [comment]
### Round 2

Source collection(on RMRK1 this time): 
https://deploy-preview-8437--koda-canary.netlify.app/rmrk/collection/7EA1DCF47E98A25067-TREE_FACED



destination: https://deploy-preview-8437--koda-canary.netlify.app/ahk/collection/304

### Whole process

https://github.com/kodadot/nft-gallery/assets/36627808/53e3e078-2caa-418f-af8e-f10d205c45df


- Consume was called twice instead of destroy at the end, is that how it works on RMRK1 ?
- there was a good 10s delay before between comprehensive scan and finalization(around 2m in the video) 

Transactions: 
https://assethub-kusama.subscan.io/tx/0x186fb005261d3553731d118774d980dea8ba8f8da9a5672be2cb1bcf263c2fa4
https://assethub-kusama.subscan.io/tx/0xb888b03a55b6c9e59d0d633ac944b05ccd5fa3b695cd401e8aa8eaa73920393c
https://kusama.subscan.io/tx/0x8751f32fc01c2cb7728bdf901b9a5db577d4fe97645833e0df145124c8853e39
https://kusama.subscan.io/tx/0x85815fbd80e764c69daec4cc8bd06f4929c9b6683bdd6bea30ddda904806c900

**Comment by roiLeo** [comment]
> is there a open issue for that? so we can implement in the future

#4612

we can filter them on ahp/ahk, not yet implemented on rmrk side

Capture d’écran 2023-12-08 à 2 16 02 PM

**Comment by prury** [comment]
> > is there a open issue for that? so we can implement in the future
> 
> #4612
> 

thank you

**Comment by exezbcz** [comment]
btw how does it get burned when, for example, not all items are migrated from, for example, ksm to ahk? 


Will the collection owner/creator have to find it manually, or will the last person migrating the item have a step = burning in the signing process?

**Comment by vikiival** [comment]
> btw how does it get burned when, for example, not all items are migrated from, for example, ksm to ahk?

You will migrate all non-burned items, no matter what :). Or maybe can you rephrase?

**Comment by exezbcz** [comment]
> > btw how does it get burned when, for example, not all items are migrated from, for example, ksm to ahk?
> 
> You will migrate all non-burned items, no matter what :). Or maybe can you rephrase?

>Will the collection owner/creator have to find it manually, or will the last person migrating the item have a step = burning in the signing process?


I mean the collection

**Comment by preschian** [comment]
> btw how does it get burned when, for example, not all items are migrated from, for example, ksm to ahk?
> 
> Will the collection owner/creator have to find it manually, or will the last person migrating the item have a step = burning in the signing process?

The collection won't be burned. Afaik, unlike on asset hub, there is no explicit error on extrinsic for rmrk

**Comment by preschian** [comment]
is it safe to merge the PR? and continue to this https://github.com/kodadot/nft-gallery/issues/8293?

File: components/migrate/steps/SignLoader3.vue [hunk 2]
```diff
@@ -73,37 +101,57 @@ const whichIcon = () => {
...
+      urlPrefix: from,
+    },
+    from,
+  )
+  updateSteps('step3-burn')
+}
+
+const burnCollection = async () => {
+  if (fromCollectionId) {
+    await transaction(
```
Comment (vikiival) [inline components/migrate/steps/SignLoader3.vue:144]: When is transaction resolved? on block / finalized ?

File: composables/useMigrate.ts [hunk 3]
```diff
@@ -123,13 +123,15 @@ export function useMigrateDeposit(
     return 12
   })
 
-  const chainItemDeposit = computed(() =>
-    parseDeposit(
+  const chainItemDeposit = computed(() => {
+    // Calculate the sum of all deposits and then multiply by itemCount squared
```
Comment (vikiival) [inline composables/useMigrate.ts:127]: Why?

Comment (preschian) [reply composables/useMigrate.ts:127]: Previously, I tried to migrate some of the items, on the deposit section shows my balance is enough. But, on the extrinsic, it failed https://assethub-kusama.subscan.io/extrinsic/0xf2759f46c9a5ac6e3e9b7f0823f7b2e245651941b32abefeab30b8b0220ecf4b

File: components/migrate/steps/SignLoader3.vue (hunk 2 above)
Comment (preschian) [reply components/migrate/steps/SignLoader3.vue:144]: mostly will check on `status.value === TransactionStatus.Finalized` and based on `steps.value`

**Comment by yangwao** [comment]
so let's merge this?

**Comment by prury** [comment]
### Round 3
Source Collection: https://deploy-preview-8437--koda-canary.netlify.app/ksm/collection/7EA1DCF47E98A25067-MUSH

image

4/5 items, @roiLeo has one NFT from this collection

Destination collection:

https://deploy-preview-8437--koda-canary.netlify.app/ahk/collection/324

whole process: 

https://github.com/kodadot/nft-gallery/assets/36627808/c8219465-9439-45c4-9b88-bb0580bdf351

txs:
https://assethub-kusama.subscan.io/tx/0x7f5a68451f79be3eb7d2589e0489be82fda5db63d202ac96a594c1a26f7c2dcd
https://assethub-kusama.subscan.io/tx/0x1fbe59caaee6cea29f988a3c30d7a3b405b177d48fc211e7a0833978def7a8d8
https://assethub-kusama.subscan.io/tx/0xe37bdc29263977f15f9f21cb13762c7accf17026343a59754a8655fb02309816
https://kusama.subscan.io/tx/0x53238d52423e8565a493e5b9c2603bb32a977f6d0b585cc4b8783caaf9a7a82f
https://kusama.subscan.io/tx/0x376d938c3458783a77896e20aa328356c237d84e8c5d1d53667f16a22fb95a45

@yangwao found some issues and added them to #8293

**Comment by vikiival** [comment]
@damskyftw can you please test?

File: components/base/MediaItem.vue [hunk 4]
```diff
@@ -100,7 +100,7 @@ const props = withDefaults(
     original: false,
     isLewd: false,
     isDetail: false,
-    placeholder: './Koda.svg',
+    placeholder: '/Koda.svg',
```
Comment (roiLeo) [inline components/base/MediaItem.vue:103]: will it work as before? or does it fix something not related with issue?

**Review by roiLeo — APPROVED** [review]
let it roll 🎢

**Comment by yangwao** [comment]
Thanks!
pay 50 usd

**Pull request merged**